### PR TITLE
Magento 2139 fav list content is synced to

### PR DESCRIPTION
**Pull request by awesselburg**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I updated the CHANGELOG.md

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,3 +9,4 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/) and this p
 - retrieval of favorite / wishlist products for guests and customers
 - addition of favorite / wishlist products for guests and customers
 - removal of favorite / wishlist products for guests and customers
+- synchronize local favorites to Magento after login / registration
```
Comment (augsteyer) [inline CHANGELOG.md]:
```suggestion
- synchronization of guest favorite / wishlist products with Magento after login / registration 
```

File: pipelines/shopgate.user.mergeFavorites.v1.json [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      },
+      {
+        "type": "extension",
+          "id": "@shopgate/magento-favorites",
+          "path": "@shopgate/magento-favorites/lib/magento/customer/getSyncItems.js",
+          "input": [
+            {"key": "sgxsMeta", "id": "10"},
+            {"key": "wishlistId", "id": "20"},
+            {"key": "mageProductsIds", "id": "30"},
+            {"key": "localProductsIds", "id": "40"}
```
Comment (augsteyer) [inline pipelines/shopgate.user.mergeFavorites.v1.json]:
```suggestion
            {"key": "guestProductIds", "id": "40"}
```

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  beforeEach(() => {
+    context = {}
+    input = {}
+  })
+
+  it('Should return correct arrays for adding one product', async () => {
+    input = {
+      mageProductsIds: [{ id: 10 }],
+      localProductsIds: [{ id: 20 }]
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]:
```suggestion
      guestProductIds: [{ id: 20 }]
```

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    const expectedAddResult = [{ id: 20 }]
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.addProductsIds, expectedAddResult)
+  })
+
+  it('Should return a empty array - same products', async () => {
+    input = {
+      mageProductsIds: [{ id: 20 }],
+      localProductsIds: [{ id: 20 }]
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]:
```suggestion
      guestProductIds: [{ id: 20 }]
```

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    const expectedAddResult = []
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.addProductsIds, expectedAddResult)
+  })
+
+  it('Should return a empty array - no local products', async () => {
+    input = {
+      mageProductsIds: [{ id: 20 }, { id: 30 }],
+      localProductsIds: []
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]:
```suggestion
      guestProductIds: []
```

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    const expectedAddResult = []
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.addProductsIds, expectedAddResult)
+  })
+
+  it('Should return correct arrays for empty mage products', async () => {
+    input = {
+      mageProductsIds: [],
+      localProductsIds: [{ id: 20 }, { id: 30 }]
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]:
```suggestion
      guestProductIds: [{ id: 20 }, { id: 30 }]
```

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    const expectedAddResult = [{ id: 20 }, { id: 30 }]
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.addProductsIds, expectedAddResult)
+  })
+
+  it('Should return correct arrays current products ids - merge', async () => {
+    input = {
+      mageProductsIds: [{ id: 40 }],
+      localProductsIds: [{ id: 20 }, { id: 30 }]
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]:
```suggestion
      guestProductIds: [{ id: 20 }, { id: 30 }]
```

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    const expectedAddResult = [{ id: 20 }, { id: 30 }]
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.productIds, expectedAddResult)
+  })
+
+  it('Should return correct arrays current products ids - no local products', async () => {
+    input = {
+      mageProductsIds: [{ id: 20 }, { id: 30 }],
+      localProductsIds: []
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]:
```suggestion
      guestProductIds: []
```

File: extension/lib/magento/customer/getSyncItems.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * @param {StepContext} context
+ * @param {Object} input
+ * @param {Object[]} input.mageProductsIds
+ * @param {Object[]} input.localProductsIds
+ * @returns {Promise<{addProducts: Object[], productIds: Object[]}>}
+ */
+module.exports = async (context, input) => {
+  const { mageProductsIds, localProductsIds } = input
```
Comment (augsteyer) [inline extension/lib/magento/customer/getSyncItems.js]:
```suggestion
  const { mageProductsIds, guestProductIds } = input
```

File: extension/lib/magento/customer/getSyncItems.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+  return {
+    addProductsIds: _.differenceWith(localProductsIds, mageProductsIds, _.isEqual),
```
Comment (augsteyer) [inline extension/lib/magento/customer/getSyncItems.js]:
```suggestion
    addProductsIds: _.differenceWith(guestProductIds, mageProductsIds, _.isEqual),
```

File: extension/lib/magento/customer/getSyncItems.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    productIds: _.concat(mageProductsIds, localProductsIds)
```
Comment (augsteyer) [inline extension/lib/magento/customer/getSyncItems.js]:
```suggestion
    productIds: _.concat(mageProductsIds, guestProductIds)
```

File: extension/lib/magento/customer/getSyncItems.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  isEqual: require('lodash/isEqual'),
+  differenceWith: require('lodash/differenceWith'),
+  concat: require('lodash/concat')
+}
+
+/**
+ * @param {StepContext} context
+ * @param {Object} input
+ * @param {Object[]} input.mageProductsIds
+ * @param {Object[]} input.localProductsIds
```
Comment (augsteyer) [inline extension/lib/magento/customer/getSyncItems.js]:
```suggestion
 * @param {Object[]} input.guestProductIds
```

File: pipelines/shopgate.user.mergeFavorites.v1.json [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+      {
+        "type": "errorCatchExtension",
+        "id": "@shopgate/favorites",
+        "path": "@shopgate/favorites/catch/catchGenericError.js",
+        "input": [],
+        "output": []
+      }
+    ]
+  }
+}
```
Comment (augsteyer) [inline pipelines/shopgate.user.mergeFavorites.v1.json]: newline

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+      localProductsIds: [{ id: 20 }, { id: 30 }]
+    }
+
+    const expectedAddResult = [{ id: 40 }, { id: 20 }, { id: 30 }]
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.productIds, expectedAddResult)
+  })
+
+  it('Should return correct arrays current products ids - no mage products', async () => {
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]: wait a second, don't you already have this test above? I mean, I know it compares different things, but you are allowed to have more than 1 assert within one test.

File: pipelines/shopgate.user.getCustomerFavorites.v1.json [hunk 15]
```diff
@@ -30,10 +31,22 @@
...
+        "output": [
+          {"key": "productIds", "id": "150"}
+        ]
       }
     ],
     "output": [
-      {"key": "productIds", "id": "120"}
+      {"key": "productIds", "id": "150"}
     ],
     "public": true
```
Comment (augsteyer) [inline pipelines/shopgate.user.getCustomerFavorites.v1.json]: can you move `public` above `input`. I think it's a better arrangement for readability. I do believe output should be at the bottom though, unlike the way we had been doing.

File: pipelines/shopgate.user.getCustomerFavorites.v1.json [hunk 16]
```diff
@@ -30,10 +31,22 @@
...
           {"key": "productIds", "id": "120"}
         ]
+      },
+      {
+        "type": "pipeline",
+        "id": "shopgate.user.mergeFavorites.v1",
+        "input": [
+          {"key": "sgxsMeta", "id": "10"},
+          {"key": "wishlistId", "id": "110"},
+          {"key": "mageProducts", "id": "120"}
```
Comment (augsteyer) [inline pipelines/shopgate.user.getCustomerFavorites.v1.json]: maybe this also can be `mageProductIds`?

File: pipelines/shopgate.user.mergeFavorites.v1.json [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+  "version": "1",
+  "pipeline": {
+    "id": "shopgate.user.mergeFavorites.v1",
+    "public": true,
+    "input": [
+      {"key": "sgxsMeta", "id": "10"},
+      {"key": "wishlistId", "id": "20"},
+      {"key": "mageProducts", "id": "30"}
+    ],
+    "output": [
```
Comment (augsteyer) [inline pipelines/shopgate.user.mergeFavorites.v1.json:11]: can u move this key below steps (at the end)

File: extension/test/newman/collection.json [hunk 18]
```diff
@@ -1530,6 +1530,1444 @@
 					"_postman_isSubFolder": true
 				}
 			]
+		},
+		{
+			"name": "Sync",
```
Comment (augsteyer) [inline extension/test/newman/collection.json]: the test structure is off, I mean, we have "guest", "customer" and "sync". I feel like something went horribly wrong with this structure ^_^ @andre-kraus , can you help? In the last fav review my suggestion was to break it down into: 
```
-> get/add/delete
-> update
-> sync
```
I think it is better, no?

File: extension/test/unit/lib/magento/customer/syncItems-test.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      localProductsIds: [{ id: 20 }, { id: 30 }]
+    }
+
+    const expectedAddResult = [{ id: 20 }, { id: 30 }]
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.productIds, expectedAddResult)
+  })
+
+  it('Should return correct arrays current products ids - no local products', async () => {
```
Comment (augsteyer) [inline extension/test/unit/lib/magento/customer/syncItems-test.js]: `Should return correct arrays current products ids` in all tests is a bit confusing. Are you trying to say `Should not return guest product ids`?

**Review by augsteyer — CHANGES_REQUESTED** [review]
also, the name of the test file is off

File: pipelines/shopgate.user.mergeFavorites.v1.json (hunk 17 above)
Comment (andre-kraus) [reply pipelines/shopgate.user.mergeFavorites.v1.json:11]: Wouldn't recommend that. If it's a pipeline with a lot of steps, you need to always scroll down to the what the pipeline is going to return. So signature should be easily visible (input, output on top).

**Review by andre-kraus — CHANGES_REQUESTED** [review]

File: extension/test/unit/lib/magento/customer/getSyncItems-test.js [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+describe('getSyncItems', () => {
+  let context
+  let input
+
+  beforeEach(() => {
+    context = {}
+    input = {}
+  })
+
+  it('Should return the full list of products ids and correct array of product id(s) to add by adding one guest product', async () => {
```
Comment (andre-kraus) [inline extension/test/unit/lib/magento/customer/getSyncItems-test.js]: Should be product ids not products ids

**Comment by augsteyer** [comment]
Alright, my environment still is not in good shape so I cannot test the app function or test running.
@andre-kraus can you check the newman tests and run the following app scenario:
1) login as user X
2) add item to favorites
3) logout
4) add items to favorites
5) login as user X
Check items merged

File: extension/test/unit/lib/magento/customer/getSyncItems-test.js [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+
+    const expectedAddResult = [20]
+    const expectedProductIdsResult = [10, 20]
+    const diffObject = await step(context, input)
+
+    assert.deepStrictEqual(diffObject.addProductsIds, expectedAddResult)
+    assert.deepStrictEqual(diffObject.productIds, expectedProductIdsResult)
+  })
+
+  it('Should return the full list of products ids and a empty array of products id(s) to add because the product products are the same', async () => {
```
Comment (andre-kraus) [inline extension/test/unit/lib/magento/customer/getSyncItems-test.js]: Same as above, please fix all occurrences.

**Review by andre-kraus — CHANGES_REQUESTED** [review]

**Comment by andre-kraus** [comment]
> Alright, my environment still is not in good shape so I cannot test the app function or test running.
> @andre-kraus can you check the newman tests and run the following app scenario:
> 
> 1. login as user X
> 2. add item to favorites
> 3. logout
> 4. add items to favorites
> 5. login as user X
>    Check items merged

Working

**Review by andre-kraus — APPROVED** [review]

**Review by augsteyer — APPROVED** [review]

**Pull request merged**